### PR TITLE
Use disk by id for mounting devices in OpenStack

### LINKED ISSUE #252: OpenStack API shows incorrect path for volume
OpenStack api shows path for volume:

`/dev/vdd on master-zone-1-2-1-ownold-master-k8s-local`

logs:

```
I0913 18:04:49.564932    1578 mounter.go:64] Master volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" is attached at "/dev/vdd"
I0913 18:04:49.564993    1578 mounter.go:78] Doing safe-format-and-mount of /dev/vdd to /mnt/master-2.etcd-main.ownold-master.k8s.local
I0913 18:04:49.565025    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:50.565341    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:51.565517    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:52.565714    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:53.565907    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:54.566067    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:55.566275    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:56.566473    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:57.566669    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:58.566883    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
I0913 18:04:59.567076    1578 mounter.go:113] Waiting for volume "44a9c2a0-4648-4ce1-8f68-34a081521ba1" to be mounted
```

However, when I check devices: 

```
$ ls /dev/vd
vda   vda1  vdb   vdc
```

So OpenStack API is reporting incorrect device paths, this is known problem from k8s side. It is solved using this function https://github.com/kubernetes/kubernetes/blob/release-1.8/pkg/cloudprovider/providers/openstack/openstack_volumes.go#L331

When listing volumes by id
```
$ ls /dev/disk/by-id
virtio-44a9c2a0-4648-4ce1-8
```
so the disk is there, but the path is incorrect. Etcd manager thinks that it is in `/dev/vdd` instead of `/dev/vdc`

### PR DESCRIPTION
**Pull request by zetaab**
fixes #252

### CODE REVIEW

**Comment by zetaab** [comment]
@justinsb if possible could you check this PR and possible compile new version available for Kops. We are seeing lots of failed clusters (because disk is not mounted) in our e2e tests (which is creating kops clusters heavily).

**Comment by zetaab** [comment]
```
root@master-zone-1-1-1-ownold-n1-k8s-local:/home/debian# docker logs -f 0884aa4a5773
etcd-manager
I0914 07:29:58.011372    8967 volumes.go:201] Found project="9ebaa4069d4d4e009a94e85a811ea8a1"
I0914 07:29:58.011445    8967 volumes.go:210] Found instanceName="master-zone-1-1-1-ownold-n1-k8s-local"
I0914 07:29:58.865072    8967 volumes.go:230] Found internalIP="10.1.32.7" and zone="zone-1"
I0914 07:29:58.865114    8967 main.go:254] Mounting available etcd volumes matching tags [KubernetesCluster=ownold-n1.k8s.local k8s.io/etcd/events k8s.io/role/master=1]; nameTag=k8s.io/etcd/events
I0914 07:29:58.865137    8967 volumes.go:300] Listing Openstack disks in 9ebaa4069d4d4e009a94e85a811ea8a1/zone-1
I0914 07:29:59.241377    8967 mounter.go:288] Trying to mount master volume: "b91a41e7-22ed-4530-bb76-0972c7883f8e"
W0914 07:30:00.648165    8967 mounter.go:293] Error attaching volume "b91a41e7-22ed-4530-bb76-0972c7883f8e": error attaching volume b91a41e7-22ed-4530-bb76-0972c7883f8e to server ed1a3aaa-890c-495a-9b2f-e335f5d908ee: Bad request with: [POST https://foo.com:13774/v2.1/servers/ed1a3aaa-890c-495a-9b2f-e335f5d908ee/os-volume_attachments], error message: {"badRequest": {"message": "Invalid input received: Invalid volume: Volume b91a41e7-22ed-4530-bb76-0972c7883f8e status must be available or downloading (HTTP 400) (Request-ID: req-eb80e0cd-f599-44ec-99e4-6912584dac98)", "code": 400}}
I0914 07:30:00.648183    8967 mounter.go:288] Trying to mount master volume: "15a67c6a-a6e6-4c33-b796-aa5dbfa04fb3"
I0914 07:30:01.883002    8967 mounter.go:302] Currently attached volumes: [0xc000245a80]
I0914 07:30:01.883022    8967 mounter.go:64] Master volume "15a67c6a-a6e6-4c33-b796-aa5dbfa04fb3" is attached at "/dev/vdc"
I0914 07:30:01.883037    8967 mounter.go:78] Doing safe-format-and-mount of /dev/vdc to /mnt/master-1.etcd-events.ownold-n1.k8s.local
W0914 07:30:01.883068    8967 volumes.go:350] Failed to find device for the volumeID: "15a67c6a-a6e6-4c33-b796-aa5dbfa04fb3"
W0914 07:30:01.883087    8967 mounter.go:82] unable to mount master volume: "failed to find device path for volume"
I0914 07:30:01.883093    8967 boot.go:49] waiting for volumes
I0914 07:31:01.883816    8967 volumes.go:300] Listing Openstack disks in 9ebaa4069d4d4e009a94e85a811ea8a1/zone-1
I0914 07:31:02.306304    8967 mounter.go:64] Master volume "15a67c6a-a6e6-4c33-b796-aa5dbfa04fb3" is attached at "/dev/vdc"
I0914 07:31:02.306327    8967 mounter.go:78] Doing safe-format-and-mount of /dev/vdc to /mnt/master-1.etcd-events.ownold-n1.k8s.local
I0914 07:31:02.306381    8967 volumes.go:344] Found disk attached as "virtio-15a67c6a-a6e6-4c33-b"; full devicepath: /rootfs/dev/disk/by-id/virtio-15a67c6a-a6e6-4c33-b
I0914 07:31:02.306391    8967 mounter.go:116] Found volume "15a67c6a-a6e6-4c33-b796-aa5dbfa04fb3" mounted at device "/dev/disk/by-id/virtio-15a67c6a-a6e6-4c33-b"
I0914 07:31:02.307108    8967 mounter.go:161] Creating mount directory "/rootfs/mnt/master-1.etcd-events.ownold-n1.k8s.local"
I0914 07:31:02.307197    8967 mounter.go:166] Mounting device "/dev/disk/by-id/virtio-15a67c6a-a6e6-4c33-b" on "/mnt/master-1.etcd-events.ownold-n1.k8s.local"
```

then after some seconds we have working etcd cluster

```
0914 07:32:51.304642    8967 controller.go:269] I am leader with token "IOPZLWMLbqLuC8rgYoKlww"
I0914 07:32:51.348641    8967 controller.go:276] etcd cluster state: etcdClusterState
  members:
    {"name":"etcd-events-1","peerURLs":["https://etcd-events-1.internal.ownold-n1.k8s.local:2381"],"endpoints":["https://etcd-events-1.internal.ownold-n1.k8s.local:4002"],"ID":"12936700238611359580"}
    {"name":"etcd-events-3","peerURLs":["https://etcd-events-3.internal.ownold-n1.k8s.local:2381"],"endpoints":["https://etcd-events-3.internal.ownold-n1.k8s.local:4002"],"ID":"17673936026932532914"}
    {"name":"etcd-events-2","peerURLs":["https://etcd-events-2.internal.ownold-n1.k8s.local:2381"],"endpoints":["https://etcd-events-2.internal.ownold-n1.k8s.local:4002"],"ID":"9072804499909938521"}
```

```
 kubectl get nodes
NAME                                    STATUS   ROLES    AGE     VERSION
master-zone-1-1-1-ownold-n1-k8s-local   Ready    master   2m11s   v1.15.3
master-zone-1-2-1-ownold-n1-k8s-local   Ready    master   2m13s   v1.15.3
master-zone-1-3-1-ownold-n1-k8s-local   Ready    master   2m10s   v1.15.3
nodes-1-ownold-n1-k8s-local             Ready    node     99s     v1.15.3
nodes-2-ownold-n1-k8s-local             Ready    node     90s     v1.15.3
```

File: pkg/volumes/openstack/volumes.go [hunk 1]
```diff
@@ -324,9 +325,38 @@ func (stack *OpenstackVolumes) FindVolumes() ([]*volumes.Volume, error) {
...
+	candidateDeviceNodes := []string{
+		// KVM
+		fmt.Sprintf("virtio-%s", volumeID[:20]),
+		// KVM virtio-scsi
+		fmt.Sprintf("scsi-0QEMU_QEMU_HARDDISK_%s", volumeID[:20]),
+		// ESXi
+		fmt.Sprintf("wwn-0x%s", strings.Replace(volumeID, "-", "", -1)),
+	}
+
+	files, _ := ioutil.ReadDir(volumes.PathFor("/dev/disk/by-id/"))
```
Comment (justinsb) [inline pkg/volumes/openstack/volumes.go]: Nit: we probably should at least log the error if there is one here, we may want to just return immediately (I'd expect files to be empty)

File: pkg/volumes/openstack/volumes.go [hunk 2]
```diff
@@ -324,9 +325,38 @@ func (stack *OpenstackVolumes) FindVolumes() ([]*volumes.Volume, error) {
...
+	for _, f := range files {
+		for _, c := range candidateDeviceNodes {
+			if c == f.Name() {
+				glog.V(4).Infof("Found disk attached as %q; full devicepath: %s\n", f.Name(), path.Join(volumes.PathFor("/dev/disk/by-id/"), f.Name()))
+				return path.Join("/dev/disk/by-id/", f.Name())
+			}
+		}
+	}
+
+	glog.Warningf("Failed to find device for the volumeID: %q\n", volumeID)
```
Comment (justinsb) [inline pkg/volumes/openstack/volumes.go]: Nit: no need for the trailing "\n" when using glog (or klog)

File: pkg/volumes/openstack/volumes.go [hunk 3]
```diff
@@ -324,9 +325,38 @@ func (stack *OpenstackVolumes) FindVolumes() ([]*volumes.Volume, error) {
...
+	return ""
+}
+
 // FindMountedVolume implements Volumes::FindMountedVolume
 func (_ *OpenstackVolumes) FindMountedVolume(volume *volumes.Volume) (string, error) {
-	device := volume.LocalDevice
+	device := GetDevicePath(volume.ProviderID)
```
Comment (justinsb) [inline pkg/volumes/openstack/volumes.go]: Returning (string, error) from GetDevicePath would be nice here because then we can propagate it up.

**Comment by justinsb** [comment]
Thanks @zetaab & sorry for the delay.  Some minor things here but we can probably fix at a later date.

**Pull request merged**